autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py

```diff
@@ -0,0 +1,234 @@
+            epsilon (float): The perturbation magnitude.
+
+        """
+        super().__init__(random_state=random_state,
+                         weighted_loss=weighted_loss,
+                         use_stochastic_weight_averaging=use_stochastic_weight_averaging,
+                         use_snapshot_ensemble=use_snapshot_ensemble,
+                         se_lastk=se_lastk,
+                         use_lookahead_optimizer=use_lookahead_optimizer,
+                         **lookahead_config)
+        self.epsilon = epsilon
+
+    def data_preparation(self, X: np.ndarray, y: np.ndarray,
+                         ) -> Tuple[Tuple[np.ndarray, np.ndarray], Dict[str, np.ndarray]]:
+        """Generate adversarial examples from the original inputs.
+
+        Args:
+            X (np.ndarray): The batch training features
+            y (np.ndarray): The batch training labels
+
+        Returns:
+            typing.Tuple[np.ndarray, np.ndarray]: original examples, adversarial examples.
+            typing.Dict[str, np.ndarray]: arguments to the criterion function.
+        """
+        X_adversarial = self.fgsm_attack(X, y)
+        return (X, X_adversarial), {'y_a': y}
+
+    def criterion_preparation(self, y_a: np.ndarray, y_b: np.ndarray = None, lam: float = 1.0
+                              ) -> Callable:
+        # Initial implementation, consider the adversarial loss and the normal network loss
+        # equally.
+        return lambda criterion, pred, adversarial_pred: 0.5 * criterion(pred, y_a) + \
+            0.5 * criterion(adversarial_pred, y_a)
+
+    def train_step(self, data: np.ndarray, targets: np.ndarray) -> Tuple[float, torch.Tensor]:
+        """
+        Allows to train 1 step of gradient descent, given a batch of train/labels
+
+        Args:
+            data (np.ndarray): input features to the network
+            targets (np.ndarray): ground truth to calculate loss
+
+        Returns:
+            torch.Tensor: The predictions of the network
+            float: the loss incurred in the prediction
+        """
+        # prepare
+        data = data.float().to(self.device)
+        targets = targets.long().to(self.device)
+
+        data, criterion_kwargs = self.data_preparation(data, targets)
+        original_data = data[0]
+        adversarial_data = data[1]
+
+        original_data = torch.autograd.Variable(original_data)
+        adversarial_data = torch.autograd.Variable(adversarial_data)
+
+        # training
+        self.optimizer.zero_grad()
+        original_outputs = self.model(original_data)
+        adversarial_output = self.model(adversarial_data)
+
+        loss_func = self.criterion_preparation(**criterion_kwargs)
+        loss = loss_func(self.criterion, original_outputs, adversarial_output)
+        loss.backward()
+        self.optimizer.step()
+        if self.scheduler:
+            if 'ReduceLROnPlateau' in self.scheduler.__class__.__name__:
+                self.scheduler.step(loss)
```

Comment:
As we are not using the ReduceLROnPlateau in our experiments e never really bothered to fix this correctly. but as you may remember we have a fix for this on development branch, so it will be fixed after rebase